rust-src/bulletproofs/src/set_membership_proof.rs

```diff
@@ -0,0 +1,960 @@
+use crate::{inner_product_proof::*, utils::*};
+use crypto_common::*;
+use crypto_common_derive::*;
+use curve_arithmetic::{multiexp, multiexp_table, multiexp_worker_given_table, Curve};
+use ff::Field;
+use pedersen_scheme::*;
+use rand::*;
+use random_oracle::RandomOracle;
+use std::iter::once;
+
+#[derive(Clone, Serialize, SerdeBase16Serialize, Debug)]
+#[allow(non_snake_case)]
+pub struct SetMembershipProof<C: Curve> {
+    /// Commitment to the evaluation of the indicator function I_{v} on the_set
+    pub A:        C,
+    /// Commitment to the blinding factors in s_L and s_R
+    pub S:        C,
+    /// Commitment to the t_1 coefficient of polynomial t(x)
+    pub T_1:      C,
+    /// Commitment to the t_2 coefficient of polynomial t(x)
+    pub T_2:      C,
+    /// Evaluation of t(x) at the challenge point x
+    pub tx:       C::Scalar,
+    /// Blinding factor for the commitment to tx
+    pub tx_tilde: C::Scalar,
+    /// Blinding factor for the commitment to the inner-product arguments
+    pub e_tilde:  C::Scalar,
+    /// Inner product proof
+    pub ip_proof: InnerProductProof<C>,
+}
+
+/// Error messages detailing why proof generation failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum ProverError {
+    /// The set must have a size of a power of two
+    SetSizeNotPowerOfTwo,
+    /// The length of G_H was less than |S|, which is too small
+    NotEnoughGenerators,
+    /// Could not compute the indicator
+    CouldNotFindValueInSet,
+    /// Could not generate inner product proof
+    InnerProductProofFailure,
+    /// Could not invert y
+    DivisionError,
+}
+
+/// This function takes a set (as a vector) and a value v as input.
+/// If v in S the function computes bit vectors aL and aR where
+/// aL_i = 1 <=> s_i = v
+/// and a_R is the bit-wise negation of a_L
+/// Note: For multi sets this function only sets the first hit to one.
```

Comment:
Can you document why this is the desired behaviour?

Reply:
Done.